File: app/models/feedback.py
from datetime import datetime, timezone
from enum import Enum
from typing import Annotated, Optional
from beanie import Document, Indexed
from pydantic import Field
from app.schemas.role import UserRole


class FeedbackStatus(str, Enum):
    NEW = "new"
    IN_REVIEW = "in_review"
    RESOLVED = "resolved"
    ARCHIVED = "archived"


class FeedbackCategory(str, Enum):
    APP_EXPERIENCE = "app_experience"
    DELIVERY_SERVICE = "delivery_service"
    PRICING = "pricing"
    FEATURE_REQUEST = "feature_request"
    BUG_REPORT = "bug_report"
    SUPPORT = "support"
    GENERAL = "general"


class Feedback(Document):
    """MongoDB Document model for App Feedback submitted by Customers and Delivery Partners."""

    user_id: Annotated[str, Indexed()]
    user_name: str
    user_email: str
    user_phone: Optional[str] = None
    role: UserRole = UserRole.USER
    rating: float = Field(..., ge=1.0, le=5.0, description="App satisfaction rating (1.0 to 5.0)")
    category: FeedbackCategory = FeedbackCategory.GENERAL
    title: Optional[str] = Field(None, max_length=200, description="Optional feedback summary or title")
    message: str = Field(..., min_length=3, max_length=2000, description="Detailed feedback or suggestion")

    # Optional client / device telemetry
    app_version: Optional[str] = None
    device_os: Optional[str] = None
    device_model: Optional[str] = None

    # Admin triage & resolution
    status: FeedbackStatus = FeedbackStatus.NEW
    admin_notes: Optional[str] = None
    admin_response: Optional[str] = None

    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    class Settings:
        name = "feedbacks"
        indexes = [
            "user_id",
            "role",
            "status",
            "category",
            [("role", 1), ("created_at", -1)],
            [("status", 1), ("created_at", -1)],
            [("rating", -1)],
            [("created_at", -1)],
        ]

    def touch(self):
        """Update updated_at timestamp."""
        self.updated_at = datetime.now(timezone.utc)

File: src/models/Feedback.ts
import mongoose, { Schema, Document } from 'mongoose';
import { UserRole } from './User';

export enum FeedbackStatus {
  NEW = 'new',
  IN_REVIEW = 'in_review',
  RESOLVED = 'resolved',
  ARCHIVED = 'archived',
}

export enum FeedbackCategory {
  APP_EXPERIENCE = 'app_experience',
  DELIVERY_SERVICE = 'delivery_service',
  PRICING = 'pricing',
  FEATURE_REQUEST = 'feature_request',
  BUG_REPORT = 'bug_report',
  SUPPORT = 'support',
  GENERAL = 'general',
  UI_UX = 'ui_ux',
  DELIVERY_SPEED = 'delivery_speed',
  DELIVERY_BEHAVIOR = 'delivery_behavior',
  APP_BUG = 'app_bug',
}

export interface IFeedback extends Document {
  user_id: string;
  user_name: string;
  user_email: string;
  user_phone?: string;
  role: UserRole;
  rating: number;
  category: FeedbackCategory;
  title?: string;
  message: string;
  app_version?: string;
  device_os?: string;
  device_model?: string;
  status: FeedbackStatus;
  admin_notes?: string;
  admin_response?: string;
  created_at: Date;
  updated_at: Date;
  touch(): void;
}

const FeedbackSchema = new Schema<IFeedback>(
  {
    user_id: { type: String, required: true, index: true },
    user_name: { type: String, required: true },
    user_email: { type: String, required: true },
    user_phone: { type: String },
    role: { type: String, enum: Object.values(UserRole), default: UserRole.USER, index: true },
    rating: { type: Number, required: true, min: 1, max: 5 },
    category: { type: String, enum: Object.values(FeedbackCategory), default: FeedbackCategory.GENERAL, index: true },
    title: { type: String, maxlength: 200 },
    message: { type: String, required: true, minlength: 3, maxlength: 2000 },
    app_version: { type: String },
    device_os: { type: String },
    device_model: { type: String },
    status: { type: String, enum: Object.values(FeedbackStatus), default: FeedbackStatus.NEW, index: true },
    admin_notes: { type: String },
    admin_response: { type: String },
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
  }
);

FeedbackSchema.index({ role: 1, created_at: -1 });
FeedbackSchema.index({ status: 1, created_at: -1 });
FeedbackSchema.index({ rating: -1 });
FeedbackSchema.index({ created_at: -1 });

FeedbackSchema.methods.touch = function () {
  this.updated_at = new Date();
};

export const Feedback = mongoose.model<IFeedback>('Feedback', FeedbackSchema, 'feedbacks');
